mava/systems/q_learning/rec_iql.py

```diff
@@ -0,0 +1,651 @@
+
+        # Next obs and term_or_trunc for learner state
+        next_obs = next_timestep.observation
+        # make compatible with network input and transition storage in next step
+        next_terminal = (1 - next_timestep.discount[..., 0, jnp.newaxis]).astype(bool)
+        next_term_or_trunc = next_timestep.last()[..., jnp.newaxis]
+
+        # Repack
+        new_act_state = ActionState(
+            next_action_selection_state,
+            next_env_state,
+            next_buffer_state,
+            next_obs,
+            next_terminal,
+            next_term_or_trunc,
+        )
+
+        return new_act_state, next_timestep.extras["episode_metrics"]
+
+    # Training functions
+
+    def prep_inputs_to_scannedrnn(obs: Observation, term_or_trunc: chex.Array) -> chex.Array:
+        """
+        Prepares the inputs to the RNN network for either
+        getting q values or the eps-greedy distribution.
+        """
+
+        hidden_state = ScannedRNN.initialize_carry(
+            (cfg.system.sample_batch_size, obs.agents_view.shape[2]), cfg.network.hidden_state_dim
+        )
+        # the rb outputs (B, T, ... ) the RNN takes in (T, B, ...)
+        obs = switch_leading_axes(obs)  # (B, T) -> (T, B)
+        term_or_trunc = switch_leading_axes(term_or_trunc)  # (B, T) -> (T, B)
+        obs_term_or_trunc = (obs, term_or_trunc)
+
+        return hidden_state, obs_term_or_trunc
+
+    def q_loss_fn(
+        q_online_params: FrozenVariableDict,
+        obs: Array,
+        term_or_trunc: Array,
+        action: Array,
+        target: Array,
+    ) -> Tuple[Array, Metrics]:
+
+        # axes switched here to scan over time
+        hidden_state, obs_term_or_trunc = prep_inputs_to_scannedrnn(obs, term_or_trunc)
+
+        # get online q values of all actions
+        _, q_online = q_net.apply(
+            q_online_params, hidden_state, obs_term_or_trunc, method="get_q_values"
+        )
+        q_online = switch_leading_axes(q_online)  # (T, B, ...) -> (B, T, ...)
+        # get the q values of the taken actions and remove extra dim
+        q_online = jnp.squeeze(
+            jnp.take_along_axis(q_online, action[..., jnp.newaxis], axis=-1), axis=-1
+        )
+        q_error = jnp.square(q_online - target)
+        q_loss = jnp.mean(q_error)  # mse
+
+        # pack metrics for logging
+        loss_info = {
+            "q_loss": q_loss,
+            "mean_q": jnp.mean(q_online),
+            "mean_target": jnp.mean(target),
+        }
+
+        return q_loss, loss_info
+
+    def update_q(
+        params: DDQNParams, opt_states: optax.OptState, data: Transition, t_train: int
+    ) -> Tuple[DDQNParams, optax.OptState, Metrics]:
+        """Update the Q parameters."""
+
+        # Get the data associated with obs
+        data_first = jax.tree_map(lambda x: x[:, :-1, ...], data)
+        # Get the data associated with next_obs
+        data_next = jax.tree_map(lambda x: x[:, 1:, ...], data)
```

Comment:
comments where no longer true because of next_obs change
```suggestion
        # Get data aligned with current/next timestep
        data_first = jax.tree_map(lambda x: x[:, :-1, ...], data)
        data_next = jax.tree_map(lambda x: x[:, 1:, ...], data)
```